From 9e8a2260945a12043f6976e4e72d0161455ff0b9 Mon Sep 17 00:00:00 2001
From: Rob Clark <rclark@redhat.com>
Date: Wed, 17 Dec 2014 21:38:25 -0500
Subject: [drm] radeon/atom: add new voltage fetch function for hawaii

Message-id: <1418852380-13061-15-git-send-email-rclark@redhat.com>
Patchwork-id: 102145
O-Subject: [RHEL7 drm 14/89] drm/radeon/atom: add new voltage fetch function for hawaii
Bugzilla: 1173317
RH-Acked-by: Jerome Glisse <jglisse@redhat.com>
RH-Acked-by: Dave Airlie <airlied@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1173317
Upstream: since v3.16.4

commit d528dcb9cac01b5ced231134e1efc7e88b4d9b20

Author:     Alex Deucher <alexander.deucher@amd.com>
AuthorDate: Thu Jul 31 17:57:42 2014 -0400
Commit:     Rob Clark <rclark@redhat.com>
CommitDate: Wed Dec 10 17:08:44 2014 -0500

    drm/radeon/atom: add new voltage fetch function for hawaii

    commit e9f274b2a1bd4ecc569b823b1e7942e9bf92593e upstream.

    Some hawaii boards use a different method for fetching the
    voltage information from the vbios.

    Signed-off-by: Alex Deucher <alexander.deucher@amd.com>
    Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>

Signed-off-by: Rob Clark <rclark@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/gpu/drm/radeon/radeon.h b/drivers/gpu/drm/radeon/radeon.h
index 60c47f8..2d6b55d 100644
--- a/drivers/gpu/drm/radeon/radeon.h
+++ b/drivers/gpu/drm/radeon/radeon.h
@@ -304,6 +304,9 @@ int radeon_atom_get_leakage_vddc_based_on_leakage_params(struct radeon_device *r
         u16 *vddc, u16 *vddci,
         u16 virtual_voltage_id,
         u16 vbios_voltage_id);
+int radeon_atom_get_voltage_evv(struct radeon_device *rdev,
+    u16 virtual_voltage_id,
+    u16 *voltage);
 int radeon_atom_round_to_true_voltage(struct radeon_device *rdev,
           u8 voltage_type,
           u16 nominal_voltage,
diff --git a/drivers/gpu/drm/radeon/radeon_atombios.c b/drivers/gpu/drm/radeon/radeon_atombios.c
index d5328f2..5f19e52 100644
--- a/drivers/gpu/drm/radeon/radeon_atombios.c
+++ b/drivers/gpu/drm/radeon/radeon_atombios.c
@@ -3236,6 +3236,41 @@ int radeon_atom_get_leakage_vddc_based_on_leakage_params(struct radeon_device *r
  return 0;
 }
 
+union get_voltage_info {
+ struct  _GET_VOLTAGE_INFO_INPUT_PARAMETER_V1_2 in;
+ struct  _GET_EVV_VOLTAGE_INFO_OUTPUT_PARAMETER_V1_2 evv_out;
+};
+
+int radeon_atom_get_voltage_evv(struct radeon_device *rdev,
+    u16 virtual_voltage_id,
+    u16 *voltage)
+{
+ int index = GetIndexIntoMasterTable(COMMAND, GetVoltageInfo);
+ u32 entry_id;
+ u32 count = rdev->pm.dpm.dyn_state.vddc_dependency_on_sclk.count;
+ union get_voltage_info args;
+
+ for (entry_id = 0; entry_id < count; entry_id++) {
+  if (rdev->pm.dpm.dyn_state.vddc_dependency_on_sclk.entries[entry_id].v ==
+      virtual_voltage_id)
+   break;
+ }
+
+ if (entry_id >= count)
+  return -EINVAL;
+
+ args.in.ucVoltageType = VOLTAGE_TYPE_VDDC;
+ args.in.ucVoltageMode = ATOM_GET_VOLTAGE_EVV_VOLTAGE;
+ args.in.ulSCLKFreq =
+  cpu_to_le32(rdev->pm.dpm.dyn_state.vddc_dependency_on_sclk.entries[entry_id].clk);
+
+ atom_execute_table(rdev->mode_info.atom_context, index, (uint32_t *)&args);
+
+ *voltage = le16_to_cpu(args.evv_out.usVoltageLevel);
+
+ return 0;
+}
+
 int radeon_atom_get_voltage_gpio_settings(struct radeon_device *rdev,
        u16 voltage_level, u8 voltage_type,
        u32 *gpio_value, u32 *gpio_mask)
-- 
1.7.1